네트워크에 필요한 VO 구조체 작성

z3rosmith opened the pull request:
- Ticker VO 작성
  - Ticker API에서 All 옵션을 사용했을 때 이용되는 ValueObject를 구현했습니다
  - QuantumValue를 이용해서 여러 타입을 한번에 decode 할수있도록 하였습니다
- Orderbook VO 작성
- Transaction VO 작성
- AssetStatus VO 작성

z3rosmith committed:
feat: TickerAllValueObject 구현
- Ticker API에서 All 옵션을 사용했을 때 이용되는 ValueObject를
  구현했습니다
- QuantumValue를 이용해서 여러 타입을 한번에 decode 할수있도록
  하였습니다

z3rosmith committed:
fea: TickerOneValueObject 구현
- 코인 상세화면에 사용되는 TickerOneValueObject를 구현하였습니다.

z3rosmith committed:
feat: Value Object 구현
- Orderbook VO 작성
- Transaction VO 작성
- AssetStatus VO 작성

z3rosmith committed:
refactor: switch문을 if case let문으로 변경
- switch에서 한 case만 쓰기때문에 switch문보다는 if case let이
  나아보였습니다.
- switch case let에서 if case let문으로 변경하였습니다.

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/OrderbookOneValueObject.swift:
```diff
@@ -0,0 +1,25 @@
+
+import Foundation
+
+struct OrderbookOneValueObject: Decodable {
+    let status: String
+    let data: OrderbookData
+}
+
+struct OrderbookData: Decodable {
+    let timestamp: String
+    let paymentCurrency: String
+    let orderCurrency: String
+    let bids: [BidData]
+    
+    struct BidData: Decodable {
+        let price: String
+        let quantity: String
+    }
+}
```
OrderbookOneValueObject 나 OrderbookData 보다, OneOrderbook, Orderbook은 어떠신가요? 아예 VO나 Data와 같은 접미어를 다 빼는 것도 의미전달이 되지 않을까 싶습니다.

z3rosmith replied:
ValueObject, Data를 빼면 언제 쓰이는지 추론하기 힘들어지지 않을까요?

donggeonoh replied:
이미 Orderbook과 같이 앞에서 추론이 된다고 생각이되긴 하는데요. 허황 의견은 어떠신가요?
@HJEHA

HJEHA replied:
회의실에서 의논 완료했습니다.

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/TickerAllValueObject.swift:
```diff
@@ -0,0 +1,83 @@
+    let data: [String: QuantumValue]
+}
+
+enum QuantumValue: Decodable {
```
QuantumValue 보다 좀 더 구체적인 네이밍이 있을까요?
같이 고민해보는게 좋을 것 같아요!

z3rosmith replied:
혹시 생각하시는 네이밍 있을까요?

donggeonoh replied:
사실 생각나는 네이밍은 없네요 ㅠㅠ PairValue는 어떨까요?

HJEHA replied:
String, TickerData 타입을 유연하게 할당해주는 것 같아서 `유연한~`이 들어가면 좋을 것 같습니다.

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/TickerAllValueObject.swift:
```diff
@@ -0,0 +1,83 @@
+}
+
+enum QuantumValue: Decodable {
+    case string(String), tickerData(TickerData)
```
case 문은 분리해주는게 좋을 것 같습니다~

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/TickerAllValueObject.swift:
```diff
@@ -0,0 +1,83 @@
+        throw QuantumError.missingValue
+    }
+    
+    enum QuantumError:Error {
```
QuantumError: Error 로
한칸 띄워쓰셔야 할 것 같습니다~

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/TickerAllValueObject.swift:
```diff
@@ -0,0 +1,83 @@
+    }
+    
+    enum QuantumError:Error {
+        case missingValue
```
에러의 경우 해당 타입으로 디코딩이 되지 않는 경우니까, couldNotTransform 이 어떠신가요?

z3rosmith replied:
저도 missingValue는 정확하지 않은 의미인것같네요.
그런데 String도 아니고 TickerData도 아닌 경우니까 noMatchingType은 어떨까요?

donggeonoh replied:
그것도 좋은 생각입니다~ noMatchingType 의미 전달에 제대로 돼서 좋네요 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/TickerAllValueObject.swift:
```diff
@@ -0,0 +1,83 @@
+    let fluctateRate24Hour: String
+    let date: String?
+
+    enum CodingKeys: String, CodingKey {
+        case openingPrice
+        case closingPrice
+        case minPrice
+        case maxPrice
+        case unitsTraded
+        case accTradeValue
+        case prevClosingPrice
+        case date
+        
+        case unitsTraded24Hour = "unitsTraded24H"
+        case accTradeValue24Hour = "accTradeValue24H"
+        case fluctate24Hour = "fluctate24H"
+        case fluctateRate24Hour = "fluctateRate24H"
+    }
```
저는 여기서 case문 사이를 붙여도 좋을 것 같습니다.

z3rosmith replied:
이건 허황이 그룹메시지에 언급한 Apple 공식문서를 참고한건데, https://developer.apple.com/documentation/foundation/archives_and_serialization/encoding_and_decoding_custom_types
여기에 제가 작성한 식으로 되어있긴 합니다. 사실 저도 안 띄어도 좋을거같긴 한데 처음 의견 주신 @HJEHA 께서는 어떻게 생각하세요?

HJEHA replied:
[Encoding and Decoding Custom Types](https://developer.apple.com/documentation/foundation/archives_and_serialization/encoding_and_decoding_custom_types) 문서 중간 `Choose Properties to Encode and Decode Using Coding Keys` 파트에서 원시값으로 문자열을 가지는 케이스와 아닌 케이스를 한줄 띄어서 작성되어있습니다. 

혹시 문서를 보고 저렇게 작성하신 것 같은데 어떻게 처리할 지 의논해 봐도 괜찮을까요??

z3rosmith replied:
저는 오른쪽에 String rawValue들이 들어가있어서 줄바꿈이 없어도 구분하기가 쉬워서 case문 사이를 안 띄워도 괜찮을것같습니다.

donggeonoh replied:
저도 위 제로 의견에 동의합니다. 붙여도 가독성에 문제 없어보이긴 합니다.

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/TickerAllValueObject.swift:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+extension QuantumValue {
+    
+    var tickerData: TickerData? {
+        if case let .tickerData(tickerData) = self {
+            return tickerData
+        }
+        return nil
+    }
+    
+    var string: String? {
+        if case let .string(string) = self {
+            return string
+        }
+        return nil
+    }
+}
```
혹시 여기 있는 연산 프로퍼티의 경우 데이터를 가져올 때 해당 ticker 혹은 string으로 접근 해야하는 걸까요?
그리고 string보단 date가 좋을 것 같네요 ㅎㅎ

donggeonoh reviewed (changes requested):
아무래도 VO나 Data 같은 접미사를 빼는 방법도 고민해보면 좋을 것 같아요. 어떻게 생각하시나요~~

HJEHA commented on BithumbYagomAcamedy/Model/ValueObject/OrderbookOneValueObject.swift:
```diff
@@ -0,0 +1,25 @@
+    let paymentCurrency: String
+    let orderCurrency: String
+    let bids: [BidData]
+    
```
Orderbook API 응답을 보면 `asks`가 존재하는데 프로퍼티로 없어도 괜찮은건가요???

``` json
{
    "status": "0000",
    "data": {
        "timestamp": "1645710805716",
        "payment_currency": "KRW",
        "order_currency": "BTC",
        "bids": [
            {
                "price": "43848000",
                "quantity": "0.0374"
            }
        ],
        "asks": [
            {
                "price": "43854000",
                "quantity": "0.02"
            }
        ]
    }
}
```

z3rosmith replied:
아 제가 놓쳤네요 ㅠㅠ 수정하겠습니다

HJEHA reviewed (commented):
코멘트 남겼습니다:) 확인 부탁드립니다😊

z3rosmith committed:
refactor: PR의견 반영
- TickerAllValueObject -> TickersValueObject로 이름 수정
- TickerOneValueObject -> TickerValueObject로 이름 수정
- OrderbookOneValueObject -> OrderbookValueObject로 이름 수정
- AssetStatusAllValueObject -> AssetsStatusValueObject로 이름 수정
- QuantumValue -> DynamicValue로 이름 수정
- 각각 ValueObject의 data 프로퍼티를 각각 ValueObject의 이름으로 변경
- missingValue 에러 이름을 noMatchingType으로 변경

donggeonoh reviewed (approved):
좋습니다 ㅎㅎ 근데 dateString이라고 한 이유가 있을까요?

z3rosmith commented:
@DonggeonOh 스위프트에 date가 Date타입일수도 있으니 헷갈릴수도 있다고 판단하여 dateString이라 이름지었습니다.
